staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -137,159 +199,83 @@ func New(opts OIDCOptions) (*OIDCAuthenticator, error) {
 	}
 
-	client, err := a.client()
-	if err != nil {
-		return nil, false, err
-	}
-	if err := client.VerifyJWT(jwt); err != nil {
-		return nil, false, err
-	}
-	claims, err := jwt.Claims()
+	idToken, err := verifier.Verify(ctx, token)
 	if err != nil {
-		return nil, false, err
+		return nil, false, fmt.Errorf("oidc: verify token: %v", err)
 	}
-	return a.parseTokenClaims(claims)
-}
 
-// parseTokenClaims maps a set of claims to a user. It performs basic validation such as
-// ensuring the email is verified.
-func (a *OIDCAuthenticator) parseTokenClaims(claims jose.Claims) (user.Info, bool, error) {
-	username, ok, err := claims.StringClaim(a.usernameClaim)
-	if err != nil {
-		return nil, false, err
+	var c claims
+	if err := idToken.Claims(&c); err != nil {
+		return nil, false, fmt.Errorf("oidc: parse claims: %v", err)
 	}
-	if !ok {
-		return nil, false, fmt.Errorf("cannot find %q in JWT claims", a.usernameClaim)
+	var username string
+	if err := c.unmarshalClaim(a.usernameClaim, &username); err != nil {
+		return nil, false, fmt.Errorf("oidc: parse username claims %q: %v", a.usernameClaim, err)
 	}
 
 	if a.usernameClaim == "email" {
-		verified, ok := claims["email_verified"]
-		if !ok {
-			return nil, false, errors.New("'email_verified' claim not present")
-		}
-
-		emailVerified, ok := verified.(bool)
-		if !ok {
-			// OpenID Connect spec defines 'email_verified' as a boolean. For now, be a pain and error if
-			// it's a different type. If there are enough misbehaving providers we can relax this latter.
-			//
-			// See: https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
-			return nil, false, fmt.Errorf("malformed claim 'email_verified', expected boolean got %T", verified)
+		var emailVerified bool
+		if err := c.unmarshalClaim("email_verified", &emailVerified); err != nil {
+			return nil, false, fmt.Errorf("oidc: parse 'email_verified' claim: %v", err)
 		}
-
 		if !emailVerified {
-			return nil, false, errors.New("email not verified")
+			return nil, false, fmt.Errorf("oidc: email not verified")
 		}
 	}
 
 	if a.usernamePrefix != "" {
 		username = a.usernamePrefix + username
 	}
 
-	// TODO(yifan): Add UID, also populate the issuer to upper layer.
 	info := &user.DefaultInfo{Name: username}
-
 	if a.groupsClaim != "" {
-		groups, found, err := claims.StringsClaim(a.groupsClaim)
-		if err != nil {
-			// Groups type is present but is not an array of strings, try to decode as a string.
-			group, _, err := claims.StringClaim(a.groupsClaim)
-			if err != nil {
-				// Custom claim is present, but isn't an array of strings or a string.
-				return nil, false, fmt.Errorf("custom group claim contains invalid type: %T", claims[a.groupsClaim])
+		if _, ok := c[a.groupsClaim]; ok {
+			var groups []string
+			if err := c.unmarshalClaim(a.groupsClaim, &groups); err != nil {
+				var group string
+				if err := c.unmarshalClaim(a.groupsClaim, &group); err != nil {
```

Comment:
A comment about why you are trying to decode it as string again as opposed to an array of strings would be helpful

Reply:
Added a link to https://github.com/kubernetes/kubernetes/issues/33290

Reply:
alternatively create a:
```
type stringOrArray []string
```
with a UnmarshalJSON func that does the right thing.